osu.Game/Screens/Select/BeatmapDetails.cs

```diff
@@ -0,0 +1,430 @@
+﻿// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using OpenTK;
+using OpenTK.Graphics;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Primitives;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Database;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+
+namespace osu.Game.Screens.Select
+{
+    public class BeatmapDetails : Container
+    {
+        private readonly MetadataSegment description;
+        private readonly MetadataSegment source;
+        private readonly MetadataSegment tags;
+
+        private readonly DifficultyRow circleSize;
+        private readonly DifficultyRow drainRate;
+        private readonly DifficultyRow overallDifficulty;
+        private readonly DifficultyRow approachRate;
+        private readonly DifficultyRow stars;
+
+        private readonly Container ratingsContainer;
+        private readonly Bar ratingsBar;
+        private readonly OsuSpriteText negativeRatings;
+        private readonly OsuSpriteText positiveRatings;
+        private readonly BarGraph ratingsGraph;
+
+        private readonly FillFlowContainer retryAndFailContainer;
+        private readonly BarGraph retryGraph;
+        private readonly BarGraph failGraph;
+
+        private BeatmapInfo beatmap;
+        public BeatmapInfo Beatmap
+        {
+            get
+            {
+                return beatmap;
+            }
+            set
+            {
+                if (beatmap == value) return;
+                beatmap = value;
+
+                description.Text = beatmap.Version;
+                source.Text = beatmap.Metadata.Source;
+                tags.Text = beatmap.Metadata.Tags;
+
+                circleSize.Value = beatmap.Difficulty.CircleSize;
+                drainRate.Value = beatmap.Difficulty.DrainRate;
+                overallDifficulty.Value = beatmap.Difficulty.OverallDifficulty;
+                approachRate.Value = beatmap.Difficulty.ApproachRate;
+                stars.Value = (float)beatmap.StarDifficulty;
+
+
+                List<int> ratings = beatmap.Metric?.Ratings?.ToList() ?? new List<int>();

```

Comment:
these local list creations just for the sake of non-null are super ugly.
````
if (beatmap.Metric?.Ratings > 0)
{
    //show logic.
}
else
{
    //hide logic.
}
````

also if you are going to just use `.ToList()` on every `IEnumerable` why not just make everything a list? (just don't call `.ToList()` IMO)